janitor/functions/truncate_datetime.py

```diff
@@ -15,46 +17,79 @@ def truncate_datetime_dataframe(
     Calling on existing df will not alter the contents
     of said df.
 
-    Note: Truncating down to a Month or Day will yields 0s,
-    as there is no 0 month or 0 day in most datetime systems.
+    This method does not mutate the original DataFrame.
+
+    Examples:
 
-    :param df: The dataframe on which to truncate datetime.
+        >>> import pandas as pd
+        >>> import janitor
+        >>> df = pd.DataFrame({
+        ...     "foo": ["xxxx", "yyyy", "zzzz"],
+        ...     "dt": pd.date_range("2020-03-11", periods=3),
+        ... })
+        >>> df
+            foo         dt
+        0  xxxx 2020-03-11
+        1  yyyy 2020-03-12
+        2  zzzz 2020-03-13
+        >>> df.truncate_datetime_dataframe("month")
+            foo         dt
+        0  xxxx 2020-03-01
+        1  yyyy 2020-03-01
+        2  zzzz 2020-03-01
+
+    :param df: The pandas DataFrame on which to truncate datetime.
     :param datepart: Truncation precision, YEAR, MONTH, DAY,
         HOUR, MINUTE, SECOND. (String is automagically
         capitalized)
 
-    :returns: a truncated datetime object to
-        the precision specified by datepart.
+    :raises ValueError: If an invalid `datepart` precision is passed in.
+    :returns: A pandas DataFrame with all valid datetimes truncated down
+        to the specified precision.
     """
-    for i in df.columns:
-        for j in df.index:
+    ACCEPTABLE_DATEPARTS = ("YEAR", "MONTH", "DAY", "HOUR", "MINUTE", "SECOND")
+    datepart = datepart.upper()
+    if datepart not in ACCEPTABLE_DATEPARTS:
+        raise ValueError(
+            "Received an invalid `datepart` precision. "
+            f"Please enter any one of {ACCEPTABLE_DATEPARTS}."
+        )
+
+    df = df.copy()
+
+    for col in df.columns:
```

Comment:
Apologies, I know I approved ... Just couldn't hake this off, what is the purpose of iterating through the index? Isn't the same function applied to every row?

Reply:
No worries,
Yeah, it is. Are you implying we can use `.apply` here?

(I originally started this PR only with the intention to add a MWE, but then realized there were bugs and warnings being thrown, so just added in some additional fixes without really intending to optimize the code any further)

Reply:
Yea, apply or if there is a faster option.

Reply:
Ok, I've modified it to use applymap so it's much faster now (5s -> 80ms on a df with 10k timestamps), and skip the df copying if a no-op is expected (when there are no datetime columns).

Reply:
Sweeeeeet